File: elixir/test/symphony_elixir/ssh_test.exs
defmodule SymphonyElixir.SSHTest do
  use ExUnit.Case, async: false

  alias SymphonyElixir.FakeSshSupport
  alias SymphonyElixir.SSH

  test "run/3 keeps bracketed IPv6 host:port targets intact" do
    test_root =
      Path.join(System.tmp_dir!(), "symphony-ssh-ipv6-test-#{System.unique_integer([:positive])}")

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file)

    assert {:ok, {"", 0}} =
             SSH.run("root@[::1]:2200", "printf ok", stderr_to_stdout: true)

    trace = File.read!(trace_file)
    assert trace =~ "-T -p 2200 root@[::1] bash -lc"
    assert trace =~ "printf ok"
  end

  test "run/3 leaves unbracketed IPv6-style targets unchanged" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-ipv6-raw-test-#{System.unique_integer([:positive])}"
      )

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file)

    assert {:ok, {"", 0}} =
             SSH.run("::1:2200", "printf ok", stderr_to_stdout: true)

    trace = File.read!(trace_file)
    assert trace =~ "-T ::1:2200 bash -lc"
    refute trace =~ "-p 2200"
  end

  test "run/3 passes host:port targets through ssh -p" do
    test_root =
      Path.join(System.tmp_dir!(), "symphony-ssh-test-#{System.unique_integer([:positive])}")

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")
    previous_ssh_config = System.get_env("SYMPHONY_SSH_CONFIG")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      restore_env("SYMPHONY_SSH_CONFIG", previous_ssh_config)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file)
    System.put_env("SYMPHONY_SSH_CONFIG", "/tmp/symphony-test-ssh-config")

    assert {:ok, {"", 0}} =
             SSH.run("localhost:2222", "echo ready", stderr_to_stdout: true)

    trace = File.read!(trace_file)
    assert trace =~ "-F /tmp/symphony-test-ssh-config"
    assert trace =~ "-T -p 2222 localhost bash -lc"
    assert trace =~ "echo ready"
  end

  test "run/3 keeps the user prefix when parsing user@host:port targets" do
    test_root =
      Path.join(System.tmp_dir!(), "symphony-ssh-user-test-#{System.unique_integer([:positive])}")

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file)

    assert {:ok, {"", 0}} =
             SSH.run("root@127.0.0.1:2200", "printf ok", stderr_to_stdout: true)

    trace = File.read!(trace_file)
    assert trace =~ "-T -p 2200 root@127.0.0.1 bash -lc"
    assert trace =~ "printf ok"
  end

  test "run/3 returns an error when ssh is unavailable" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-missing-test-#{System.unique_integer([:positive])}"
      )

    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    File.mkdir_p!(test_root)
    System.put_env("PATH", test_root)

    assert {:error, :ssh_not_found} = SSH.run("localhost", "printf ok")
  end

  test "run/3 returns a timeout error when ssh exceeds the caller timeout" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-timeout-test-#{System.unique_integer([:positive])}"
      )

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file, """
    #!/bin/sh
    printf 'ARGV:%s\\n' "$*" >> "#{trace_file}"
    sleep 1
    exit 0
    """)

    assert {:error, {:ssh_timeout, "localhost", 10}} =
             SSH.run("localhost", "printf ok", stderr_to_stdout: true, timeout: 10)

    FakeSshSupport.wait_for_trace!(trace_file)
    trace = File.read!(trace_file)
    assert trace =~ "-T localhost bash -lc"
    assert trace =~ "printf ok"
  end

  test "start_port/3 supports binary output without line mode" do
    test_root =
      Path.join(System.tmp_dir!(), "symphony-ssh-port-test-#{System.unique_integer([:positive])}")

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")
    previous_ssh_config = System.get_env("SYMPHONY_SSH_CONFIG")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      restore_env("SYMPHONY_SSH_CONFIG", previous_ssh_config)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file, """
    #!/bin/sh
    printf 'ARGV:%s\\n' "$*" >> "#{trace_file}"
    printf 'ready\\n'
    exit 0
    """)

    System.delete_env("SYMPHONY_SSH_CONFIG")

    assert {:ok, port} = SSH.start_port("localhost", "printf ok")
    assert is_port(port)
    FakeSshSupport.wait_for_trace!(trace_file)

    trace = File.read!(trace_file)
    assert trace =~ "-T localhost bash -lc"
    refute trace =~ " -F "
  end

  test "start_port/3 supports line mode" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-line-port-test-#{System.unique_integer([:positive])}"
      )

    trace_file = Path.join(test_root, "ssh.trace")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh!(test_root, trace_file, """
    #!/bin/sh
    printf 'ARGV:%s\\n' "$*" >> "#{trace_file}"
    printf 'ready\\n'
    exit 0
    """)

    assert {:ok, port} = SSH.start_port("localhost:2222", "printf ok", line: 256)
    assert is_port(port)
    FakeSshSupport.wait_for_trace!(trace_file)

    trace = File.read!(trace_file)
    assert trace =~ "-T -p 2222 localhost bash -lc"
  end

  test "remote_shell_command/1 escapes embedded single quotes" do
    assert SSH.remote_shell_command("printf 'hello'") ==
             "bash -lc 'printf '\"'\"'hello'\"'\"''"
  end

  test "write_file/4 preserves shebang-prefixed payloads without a leading newline" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-write-file-shebang-test-#{System.unique_integer([:positive])}"
      )

    trace_file = Path.join(test_root, "ssh.trace")
    output_file = Path.join(test_root, "script.sh")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh_with_eval!(test_root, trace_file)

    payload = "#!/bin/bash\necho ready\n"

    assert :ok = SSH.write_file("localhost", output_file, payload, mode: 0o755)
    written_contents = File.read!(output_file)
    assert written_contents == payload
    assert {:ok, stat} = File.stat(output_file)
    assert :erlang.band(stat.mode, 0o777) == 0o755

    trace = File.read!(trace_file)
    assert trace =~ "printf"
    assert trace =~ "%s"
  end

  test "write_file/4 preserves delimiter-shaped payload lines without execution" do
    test_root =
      Path.join(
        System.tmp_dir!(),
        "symphony-ssh-write-file-test-#{System.unique_integer([:positive])}"
      )

    trace_file = Path.join(test_root, "ssh.trace")
    output_file = Path.join(test_root, "written.txt")
    probe_file = Path.join(test_root, "pwned")
    previous_path = System.get_env("PATH")

    on_exit(fn ->
      restore_env("PATH", previous_path)
      File.rm_rf(test_root)
    end)

    FakeSshSupport.install_fake_ssh_with_eval!(test_root, trace_file)

    payload = """
    first line
    __SYMPHONY_SSH_WRITE_PAYLOAD__
    touch #{probe_file}
    second line
    """

    assert :ok = SSH.write_file("localhost", output_file, payload, mode: 0o640)
    assert File.read!(output_file) == payload
    refute File.exists?(probe_file)
    assert {:ok, stat} = File.stat(output_file)
    assert :erlang.band(stat.mode, 0o777) == 0o640

    trace = File.read!(trace_file)
    refute trace =~ "cat <<'__SYMPHONY_SSH_WRITE_PAYLOAD__' >"
    assert trace =~ "printf"
    assert trace =~ "%s"
    assert trace =~ "__SYMPHONY_SSH_WRITE_PAYLOAD__"
  end

  defp restore_env(key, nil), do: System.delete_env(key)
  defp restore_env(key, value), do: System.put_env(key, value)
end

File: ts/packages/ssh/test/ssh.test.ts
import fs from "node:fs/promises";
import path from "node:path";
import { performance } from "node:perf_hooks";
import { setTimeout as delay } from "node:timers/promises";

// Tests mutate process.env.PATH and process.env.SYMPHONY_SSH_CONFIG to inject fake
// ssh binaries and config. This requires sequential execution (enforced by the root
// vitest.config.ts `sequence: { concurrent: false }`). The afterEach hook guarantees
// env restoration even when assertions fail mid-test.
import { afterEach, beforeEach, test } from "vitest";

import { assert } from "../../../test/assert.js";
import { tempDir, writeExecutable } from "../../../test/helpers.js";

import {
  parseSshTarget,
  remoteShellCommand,
  runSsh,
  shellEscape,
  sshArgs,
  writeRemoteFile,
} from "@symphony/ssh";

let savedEnv: { PATH: string | undefined; SYMPHONY_SSH_CONFIG: string | undefined };

beforeEach(() => {
  savedEnv = {
    PATH: process.env.PATH,
    SYMPHONY_SSH_CONFIG: process.env.SYMPHONY_SSH_CONFIG,
  };
});

afterEach(() => {
  restoreEnv("PATH", savedEnv.PATH);
  restoreEnv("SYMPHONY_SSH_CONFIG", savedEnv.SYMPHONY_SSH_CONFIG);
});

test("SSH target parsing and command args match Elixir host:port behavior", () => {
  assert.deepEqual(parseSshTarget("localhost:2222"), { destination: "localhost", port: "2222" });
  assert.deepEqual(parseSshTarget("root@127.0.0.1:2200"), {
    destination: "root@127.0.0.1",
    port: "2200",
  });
  assert.deepEqual(parseSshTarget("root@[::1]:2200"), { destination: "root@[::1]", port: "2200" });
  assert.deepEqual(parseSshTarget("::1:2200"), { destination: "::1:2200", port: null });
  assert.equal(remoteShellCommand("printf 'hello'"), "bash -lc 'printf '\"'\"'hello'\"'\"''");
  assert.deepEqual(sshArgs("localhost:2222", "echo ready"), [
    "-T",
    "-p",
    "2222",
    "localhost",
    "bash -lc 'echo ready'",
  ]);
});

test("SSH run honors SYMPHONY_SSH_CONFIG, stderr folding, missing ssh, and timeouts", async () => {
  const root = await tempDir("symphony-ts-ssh");
  const trace = path.join(root, "ssh.trace");

  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
printf 'out\\n'
printf 'err\\n' >&2
exit 7
`,
  );
  process.env.SYMPHONY_SSH_CONFIG = "/tmp/symphony-test-ssh-config";

  const result = await runSsh("localhost:2222", "echo ready", { stderrToStdout: true });
  assert.equal(result.status, 7);
  assert.equal(result.stdout, "out\nerr\n");
  assert.equal(result.stderr, "");
  const traceText = await fs.readFile(trace, "utf8");
  assert.match(traceText, /-F \/tmp\/symphony-test-ssh-config -T -p 2222 localhost bash -lc/);
  assert.match(traceText, /echo ready/);

  const emptyPath = await tempDir("symphony-ts-ssh-empty-path");
  process.env.PATH = emptyPath;
  await assert.rejects(() => runSsh("localhost", "printf ok"), /ssh_not_found/);

  process.env.PATH = savedEnv.PATH!;
  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
sleep 1
exit 0
`,
  );
  await assert.rejects(
    () => runSsh("localhost", "printf ok", { timeoutMs: 20 }),
    /ssh_timeout: localhost 20/,
  );
});

test("SSH timeout rejects near the caller deadline when a child keeps pipes open", async () => {
  const root = await tempDir("symphony-ts-ssh-timeout");
  const trace = path.join(root, "ssh.trace");

  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
node -e 'process.on("SIGTERM", () => {}); setInterval(() => {}, 1000)' &
child="$!"
printf 'CHILD:%s\\n' "$child" >> ${shellEscape(trace)}
wait "$child"
`,
  );

  const started = performance.now();
  await assert.rejects(
    () => runSsh("localhost", "printf ok", { timeoutMs: 1000 }),
    /ssh_timeout: localhost 1000/,
  );
  const elapsedMs = performance.now() - started;

  if (elapsedMs >= 1_500) throw new Error(`timeout returned after ${Math.round(elapsedMs)}ms`);

  const traceText = await fs.readFile(trace, "utf8");
  const childMatch = /^CHILD:(\d+)$/m.exec(traceText);
  if (!childMatch) throw new Error(`fake ssh child pid missing in trace: ${traceText}`);
  await waitForProcessExit(Number(childMatch[1]), 7_000);
});

test("SSH writeRemoteFile preserves payload bytes and applies mode", async () => {
  const root = await tempDir("symphony-ts-ssh-write");
  const trace = path.join(root, "ssh.trace");
  const remotePath = path.join(root, "nested", "script.sh");

  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
for arg in "$@"; do last_arg="$arg"; done
eval "$last_arg"
`,
  );
  const payload = "#!/bin/bash\necho ready\n__SYMPHONY_SSH_WRITE_PAYLOAD__\n";
  await writeRemoteFile("localhost", remotePath, payload, { mode: 0o755 });
  assert.equal(await fs.readFile(remotePath, "utf8"), payload);
  const stat = await fs.stat(remotePath);
  assert.equal(stat.mode & 0o777, 0o755);
  const traceText = await fs.readFile(trace, "utf8");
  assert.match(traceText, /printf/);
  assert.notMatch(traceText, /cat <<'__SYMPHONY_SSH_WRITE_PAYLOAD__'/);
});

test("SSH writeRemoteFile rejects unsafe string modes without executing them", async () => {
  const root = await tempDir("symphony-ts-ssh-unsafe-mode");
  const trace = path.join(root, "ssh.trace");
  const marker = path.join(root, "marker");
  const remotePath = path.join(root, "script.sh");

  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
for arg in "$@"; do last_arg="$arg"; done
eval "$last_arg"
`,
  );

  let writeError: unknown;
  try {
    await writeRemoteFile("localhost", remotePath, "echo ready\n", {
      mode: `u+x; printf pwned > ${shellEscape(marker)}`,
    });
  } catch (error) {
    writeError = error;
  }

  await assertMissing(marker, "unsafe chmod mode created marker");
  assert.match(String(writeError), /invalid_chmod_mode/);
});

test("SSH writeRemoteFile shell-quotes string modes and protects dash-leading paths", async () => {
  const root = await tempDir("symphony-ts-ssh-string-mode");
  const trace = path.join(root, "ssh.trace");
  const remotePath = "-script.sh";

  await installFakeSsh(
    root,
    trace,
    `#!/bin/sh
printf 'ARGV:%s\\n' "$*" >> ${shellEscape(trace)}
for arg in "$@"; do last_arg="$arg"; done
cd ${shellEscape(root)}
eval "$last_arg"
`,
  );

  await writeRemoteFile("localhost", remotePath, "echo ready\n", { mode: "u+x" });

  const stat = await fs.stat(path.join(root, remotePath));
  assert.equal(stat.mode & 0o777, 0o744);
  const traceText = await fs.readFile(trace, "utf8");
  assert.match(traceText, /chmod '"'"'u\+x'"'"' '"'"'\.\/-script\.sh'"'"'/);
});

async function installFakeSsh(root: string, trace: string, source: string): Promise<void> {
  const bin = path.join(root, "bin");
  await fs.mkdir(bin, { recursive: true });
  await writeExecutable(path.join(bin, "ssh"), source);
  process.env.PATH = `${bin}:${process.env.PATH ?? ""}`;
  await fs.writeFile(trace, "");
}

function restoreEnv(key: string, value: string | undefined): void {
  if (value === undefined) delete process.env[key];
  else process.env[key] = value;
}

async function waitForProcessExit(pid: number, timeoutMs: number): Promise<void> {
  const deadline = performance.now() + timeoutMs;
  while (performance.now() < deadline) {
    try {
      process.kill(pid, 0);
    } catch {
      return;
    }
    await delay(50);
  }
  throw new Error(`process ${pid} still running after ${timeoutMs}ms`);
}

async function assertMissing(filePath: string, message: string): Promise<void> {
  try {
    await fs.access(filePath);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") return;
    throw error;
  }
  throw new Error(message);
}
